test/functional/api/v2/discovery_rules_controller_test.rb

```diff
@@ -31,6 +31,23 @@ class Api::V2::DiscoveryRulesControllerTest < ActionController::TestCase
     assert_equal "Location 1", discovery_rule['locations'].first['name']
   end
 
+  test_attributes :pid => 'b8ae7a80-b9a8-4924-808c-482a2b4102c4'
```

Comment:
I'd prefer to combine this test with the second one, they only differ in specifying org which users normally use, no org (also known as any context) is edge case only used by administrators

Reply:
I politely disagree. We clearly have 2 distinct use cases here that should be tested individually. For me to understand, what is your concern?

Reply:
This is a controller test, IMHO it should test the controller responsibilities, which is: taking parameters and passing it to underlying layer (model) and generate a adequate response. I don't think there's a big value of testing two parameters separately, it means going twice through the same controller code, while the only difference is the model behavior. This should and already is tested on model level, see [here](https://github.com/ldjebran/foreman_discovery/blob/e94e8c177dfb442afafa584644db14d41e7d3d55/test/unit/discovery_rule_test.rb#L87-L93) and [here](https://github.com/ldjebran/foreman_discovery/blob/e94e8c177dfb442afafa584644db14d41e7d3d55/test/unit/discovery_rule_test.rb#L14).

Therefore, I think it's fine to write single test which passes both inputs. That means only one execution but verifying both inputs works. If one of the input does not work, from the test output, it's should be easy to tell what failed, as it checks what attributes the newly created rule actually has.

TL;DR reducing execution time, deduplicating coverage to keep tests maintainable

Reply:
@omaciel many thanks for your question
@ares many thanks for the explanation but we tend to create tests with the minimum required fields(in a predefined scenario) and the mentioned model tests also use organization and location, the second add priority also.
As this test is not covered any where will move it to model module.

Reply:
@omaciel @ares but of course also move the assertions to the second test, so we will a combined coverage this test divided between the second test for the returned attributes and model test for the validation  functionality

Reply:
sounds good, thanks

Reply:
-- Therefore, I think it's fine to write single test which passes both inputs. That means only one execution but verifying both inputs works. If one of the input does not work, from the test output, it's should be easy to tell what failed, as it checks what attributes the newly created rule actually has.

@ares I see that you have reached an agreement with @ldjebran but here is why your proposal is not considered good practice for testing: assuming `parameter1` makes the test fail, `parameter2` would not be tested (unless this framework would assure that all parameters are executed?) and therefore you'd only see half the story. If the second test should fail, it would be best to see both failures at the same time.

Reply:
For me, the most important is that the test is red, I need to start investigating since my code can't get in, when I fix the original issue, I run tests locally and see there's another failure. I understand you might be interested in "how many different tests are failing", but I'd argue you'd need to have each assertion as a separate test. From my experience, that does not work well and generate a lot of failed tests for the same cause.

Reply:
@ares I agree that one assertion per test is a good rule of thumb.

Reply:
To jump in here: the model provides a method that validates. That abstraction is called in the controller. Therefor it's sufficient to test every single case fails the validation function. Ideally this is done in separate tests so you can easily see which input actually failed. Then in the controller it's sufficient to test that the validation method is called. Testing every single input at the controller level is testing at the wrong layer of abstraction. That's why it's sufficient to have 2 test cases for the controller: one with valid input and one with invalid input. What exactly triggers the invalidation is not that important.

At least, that's my view on it.

Reply:
@ekohl  No argument that testing at the controller level should suffice. I believe that @ldjebran however found an interesting case that required the extra step...